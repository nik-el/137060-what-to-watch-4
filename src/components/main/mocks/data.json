{
  "promoItem": {
    "title": "The Grand Budapest Hotel",
    "year": 2014,
    "genre": "Comedy"
  },
  "feedItems": [
    "Fantastic Beasts",
    "Bohemian Rhapsody",
    "Macbeth",
    "Terminator",
    "Terminator 2",
    "Terminator 3"
  ]
}
