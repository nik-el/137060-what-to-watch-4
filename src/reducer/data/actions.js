const getGenresFromFilms = (films) => {
  const genres = new Set();
  films.forEach((film) => genres.add(film.genre));
  return [...genres];
};

const ActionTypeAsync = {
  FILMS_REQUEST: `FILMS_REQUEST`,
  FILMS_SUCCESS: `FILMS_SUCCESS`,
  FILMS_FAILURE: `FILMS_FAILURE`,
};
const ActionCreatorAsync = {
  filmsRequest: () => ({
    type: ActionTypeAsync.FILMS_REQUEST,
  }),
  filmsSuccess: (films) => ({
    type: ActionTypeAsync.FILMS_SUCCESS,
    payload: {
      films,
      genres: getGenresFromFilms(films)
    }
  }),
  filmsFailure: (error) => ({
    type: ActionTypeAsync.FILMS_FAILURE,
    payload: error
  })
};

export {
  ActionTypeAsync,
  ActionCreatorAsync
};
