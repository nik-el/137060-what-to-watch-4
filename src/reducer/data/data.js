import {extend} from '../../utils/extend.utils';
import {filmAdapter} from '../../adapters/films';

const getGenres = (filmItems) => {
  const genres = new Set();
  filmItems.forEach((film) => genres.add(film.genre));
  return [...genres];
};

const Operation = {
  loadFilms: () => (dispatch, getState, api) => {
    dispatch(ActionCreatorAsync.filmsRequest());
    return api.get(`/films`)
      .then((response) => {
        const adaptedFilms = response.data.map((item) => filmAdapter(item));
        dispatch(ActionCreatorAsync.filmsSuccess(adaptedFilms));
        dispatch(ActionCreator.setGenres(adaptedFilms));
      }).catch(function (error) {
        dispatch(ActionCreatorAsync.filmsFailure(error));
      });
  },
};

const initialState = {
  films: [],
  loadingFilmsData: false,
  genres: [],
};

const ActionType = {
  SET_GENRES: `SET_GENRES`,
  GET_FILMS_BY_GENRE: `GET_FILMS_BY_GENRE`
};
const ActionCreator = {
  getFilmsByGenre: (filter) => ({
    type: ActionType.GET_FILMS_BY_GENRE,
    payload: filter
  }),
  setGenres: (films) => ({
    type: ActionType.SET_GENRES,
    payload: getGenres(films)
  })
};

const ActionTypeAsync = {
  FILMS_REQUEST: `FILMS_REQUEST`,
  FILMS_SUCCESS: `FILMS_SUCCESS`,
  FILMS_FAILURE: `FILMS_FAILURE`,
};
const ActionCreatorAsync = {
  filmsRequest: () => ({
    type: ActionTypeAsync.FILMS_REQUEST,
  }),
  filmsSuccess: (films) => ({
    type: ActionTypeAsync.FILMS_SUCCESS,
    payload: films
  }),
  filmsFailure: (error) => ({
    type: ActionTypeAsync.FILMS_FAILURE,
    payload: error
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypeAsync.FILMS_REQUEST:
      return extend(state, {
        loadingFilmsData: true,
      });
    case ActionTypeAsync.FILMS_SUCCESS:
      const films = action.payload;
      return extend(state, {
        films,
        loadingFilmsData: false
      });
    case ActionTypeAsync.FILMS_FAILURE:
      return extend(state, {
        films: [],
        loadingFilmsData: false
      });
    case ActionType.GET_FILMS_BY_GENRE:
      const currentGenre = action.payload;

      if (!currentGenre) {
        return state;
      }

      return extend(state, {
        films: state.films.filter((item) => item.genre === currentGenre),
      });
    case ActionType.SET_GENRES:
      const genres = action.payload;
      return extend(state, {
        genres,
      });
    default:
      return state;
  }
};

export {reducer, ActionType, ActionCreator, Operation};
