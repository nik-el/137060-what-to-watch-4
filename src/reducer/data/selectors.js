import NameSpace from "../name-space.js";

const NAME_SPACE = NameSpace.DATA;


export const getFilms = (state) => state[NAME_SPACE].films;
export const getFilmsLoadingState = (state) => state[NAME_SPACE].loadingFilmsData;
export const getGenres = (state) => state[NAME_SPACE].genres;
export const getRestApi = (state) => state[NAME_SPACE].restApi;
