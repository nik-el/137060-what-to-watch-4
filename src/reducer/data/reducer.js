import {extend} from "../../utils/extend.utils";
import {ActionType, ActionTypeAsync} from "./actions";

const initialState = {
  films: [],
  loadingFilmsData: false,
  genres: [],
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypeAsync.FILMS_REQUEST:
      return extend(state, {
        loadingFilmsData: true,
      });
    case ActionTypeAsync.FILMS_SUCCESS:
      const films = action.payload;
      return extend(state, {
        films,
        loadingFilmsData: false
      });
    case ActionTypeAsync.FILMS_FAILURE:
      return extend(state, {
        films: [],
        loadingFilmsData: false
      });
    case ActionType.SET_GENRES:
      const genres = action.payload;
      return extend(state, {
        genres,
      });
    default:
      return state;
  }
};
