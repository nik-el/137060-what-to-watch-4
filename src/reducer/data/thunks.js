import {ActionCreatorAsync} from './actions';
import {filmAdapter} from '../../adapters/films';

const Operation = {
  loadFilms: () => (dispatch, getState, api) => {
    dispatch(ActionCreatorAsync.filmsRequest());
    return api.get(`/films`)
      .then((response) => {
        const adaptedFilms = response.data.map((item) => filmAdapter(item));
        dispatch(ActionCreatorAsync.filmsSuccess(adaptedFilms));
      }).catch(function (error) {
        dispatch(ActionCreatorAsync.filmsFailure(error));
      });
  },
  getPromoFilm: () => (dispatch, getState, api) => {
    dispatch(ActionCreatorAsync.promoRequest());
    return api.get(`/films/promo`)
      .then((response) => {
        dispatch(ActionCreatorAsync.promoSuccess(response.data));
      }).catch(function (error) {
        dispatch(ActionCreatorAsync.promoFailure(error));
      });
  },
};

export {
  Operation
};
