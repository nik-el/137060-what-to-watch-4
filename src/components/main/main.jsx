import React from 'react';
import {useSelector} from "react-redux";
import {Feed} from '../feed/feed';
import {UserBlock} from '../user-block/user-block';
import {GenresList} from '../genres-list/genres-list';
import {ShowMore} from '../show-more/show-more';
import {getFilms, getFilmsLoadingState, getRestApi} from "../../reducer/data/selectors";
import {AuthorizationStatus} from "../../reducer/user/enum";
import {getAuth, getUser} from "../../reducer/user/selectors";
import {getFeedLimit} from "../../reducer/view/selectors";
import PropTypes from "prop-types";
import {Logo} from "../logo/logo";
import {FavoriteButton} from "../favorite-button/favorite-button";

const DEFAULT_LIMIT = 8;

export const Main = ({currentFilms}) => {
  const userData = useSelector(getUser);
  const restApi = useSelector(getRestApi);

  const isFilmsLoading = useSelector(getFilmsLoadingState);
  const isAuth = (useSelector(getAuth) === AuthorizationStatus.AUTH);
  const films = useSelector(getFilms);
  const feedLimit = useSelector(getFeedLimit || DEFAULT_LIMIT);

  const promoItem = films[0];

  const canShowMore = feedLimit < currentFilms.length;

  let avatarUrl;
  if (userData) {
    avatarUrl = restApi + userData.avatar_url;
  }


  if (isFilmsLoading) {
    return <div>Загружаемся...</div>;
  } else if (!films.length) {
    return <div>Нет данных для отображения</div>;
  }

  return <>
    <section className="movie-card">
      <div className="movie-card__bg">
        <img src={promoItem.bgImage} alt="The Grand Budapest Hotel"/>
      </div>

      <h1 className="visually-hidden">WTW</h1>

      <header className="page-header movie-card__head">
        <Logo />

        <UserBlock
          isAuth={isAuth}
          avatarUrl={avatarUrl}
        />

      </header>

      <div className="movie-card__wrap">
        <div className="movie-card__info">
          <div className="movie-card__poster">
            <img src={promoItem.poster}
              alt="The Grand Budapest Hotel poster"
              width="218"
              height="327"
            />
          </div>

          <div className="movie-card__desc">
            <h2 className="movie-card__title">
              {promoItem.title}
            </h2>
            <p className="movie-card__meta">
              <span className="movie-card__genre">{promoItem.genre}</span>
              <span className="movie-card__year">{promoItem.releaseYear}</span>
            </p>

            <div className="movie-card__buttons">
              <button className="btn btn--play movie-card__button" type="button">
                <svg viewBox="0 0 19 19" width="19" height="19">
                  <use xlinkHref="#play-s"></use>
                </svg>
                <span>Play</span>
              </button>
              <FavoriteButton
                id={promoItem.id}
                isFavorite={promoItem.isFavorite}
              />
            </div>
          </div>
        </div>
      </div>
    </section>

    <div className="page-content">
      <section className="catalog">
        <h2 className="catalog__title visually-hidden">Catalog</h2>

        <GenresList />

        <Feed
          className="catalog__movies-list"
          currentFilms={currentFilms}
          limit={feedLimit}
        />

        { canShowMore && <ShowMore
          offset={feedLimit}
        /> }
      </section>

      <footer className="page-footer">

        <Logo
          alignRight
        />

        <div className="copyright">
          <p>© 2019 What to watch Ltd.</p>
        </div>
      </footer>
    </div>
  </>;
};

Main.propTypes = {
  currentFilms: PropTypes.arrayOf(PropTypes.object),
};
