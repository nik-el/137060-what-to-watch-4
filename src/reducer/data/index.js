import {ActionType, ActionCreator} from './actions';
import {reducer} from "./reducer";
import {Operation} from './thunks';

export {reducer, ActionType, ActionCreator, Operation};
