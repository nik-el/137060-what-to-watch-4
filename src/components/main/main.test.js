import React from 'react';
import renderer from 'react-test-renderer';
import {Main} from './main';

import data from './mocks/data.json';
const onCardTitleClick = () => {};

it(`Main renders correctly`, () => {
  const tree = renderer
    .create(<Main
      {...data}
      onCardTitleClick={onCardTitleClick}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

