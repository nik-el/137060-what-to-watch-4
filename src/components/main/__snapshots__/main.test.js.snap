// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders correctly 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="https://picsum.photos/id/1"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <img
            alt="User avatar"
            height="63"
            src="img/avatar.jpg"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="https://picsum.photos/id/2"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Test title 1
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              test genre 1
            </span>
            <span
              className="movie-card__year"
            />
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Test genre 1
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Test genre 2
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Test title 1"
              height="175"
              src="https://picsum.photos/id/1"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
            onClick={[Function]}
          >
            <a
              className="small-movie-card__link"
              href="/detailed/1"
              onClick={[Function]}
            >
              Test title 1
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Test title 2"
              height="175"
              src="https://picsum.photos/id/2"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
            onClick={[Function]}
          >
            <a
              className="small-movie-card__link"
              href="/detailed/2"
              onClick={[Function]}
            >
              Test title 2
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
